ScanningGadget & KnownTech handling

Metious opened the pull request:
### Changes made in this pull request

resolves #402 
fixes #466

  - Improved the `KnownTechPatcher` implementation; special thanks to @vlyon for the suggestion in #402 
  - Fixed AddAnalysisTech assignment being flipped, results in overwriting values never being implemented for existing entries
  - Added `AddRequirementForUnlock` method that just adds blueprints to a requirement tech type
  - Added a better `WithScannerEntry` overload for the `ScanningGadget` that doesn't override a techtype's scanner entry that it has no business in
  - `ScanningGadget` now always adds analysis tech to the current tech type, when the `WithAnalysisTech` method is called
  - Fixed log spams for buildables
  - Added hard-locking to make items not unlockable by default or by the `unlockall` command
  - Added a cross-game-friendly overload for `AnalysisTech`.

### Breaking changes

  - You can no longer set requirements for tech types that don't have an Analysis tech

### Deprecated methods
  - [ScanningGadget.WithScannerEntry](https://github.com/Metious/Nautilus/blob/95d47bb8423450bfd5f3b7b5a372abc849685af9/Nautilus/Assets/Gadgets/ScanningGadget.cs#L253)
  
### TODO:
- [x] Decide on how to implement the `KnownTechPatcher.RemovalTechs` efficiently

Metious committed:
KnownTechPatcher cleanup

Metious committed:
Fixed AddAnalysisTech assignment being flipped

Metious committed:
Added `AddRequirementForUnlock`

Metious committed:
Added another WithScannerEntry overload and made the previous one obsolete

The previous one was overriding the PDA scanner entry for the requirement tech type, which is against the rules of gadgets rules and very confusing.

Metious committed:
ScanningGadget now always adds the analysis tech entry to the current techtype

Metious commented:
The `KnownTechPatcher.RemovalTechs` list is only used once and only for compound techs. I'd like to know why are we removing compound techs from other unlocks and if it is necessary anymore

Metious committed:
Fixed Reinitialization not working

Metious committed:
Implement RemovalTechs and RemoveFromSpecificTechs

Metious committed:
Fixed a bug where RemoveAllAnalysisTechEntry self-targets

Fixed a bug where calling the `RemoveAllCurrentAnalysisTechEntry` would also remove the unlock for the target's analysis tech entry, resulting in unexpected behavior.

Metious committed:
Added `CreateFragment` gadget extension

Metious committed:
Idk why this is here

Metious committed:
Fixed spelling mistake

MrPurple6411 commented on Nautilus/Patchers/KnownTechPatcher.cs:
A few notes on what I am seeing here.

1. This will UTTERLY break Unknown Name Mod if Reinitialize is called even once because you're effectively redoing everything over WITHOUT actually calling the method the game calls to do the Initialize and instead straight calling ONLY your patch.

2. This will end up with MANY Duplicates in the defaulttech as every time it is reinitialized you are addrange without reseting it to empty.

this is just a very quick look into these changes which I will continue to check and test using UnknownName mod to test it with.

So far, I would like the reinitialize to call the proper game method so that any mods that patch this will be able to make their changes too.

I will also try and see if I can modify my mod to better work with this new change but at the VERY least, I think this is only partially done.

Relevent Unknown Name Patch https://github.com/SubnauticaModding/Purples-Mods/blob/development/UnkownName/Patches/KnownTech_Initialize.cs

Maybe use the KnownTech.Deinitialize and then call KnownTech.Initialize(PDAData data) instead of just doing it over in the Reinitialize when theirs already does it and more so as not to have it duplicating data and allow for other mods to patch in.

![image](https://github.com/SubnauticaModding/Nautilus/assets/7858500/a0709775-40b7-4912-9a7d-70dd56e86500)


I don't think we even need the Player.main check in the if as the Initialized check should be sufficient.

so it could be as simple as: 
```cs

    internal static void Reinitialize()
    {
        // At this point the KnownTech stuff haven't been initialized yet, so no need to re-patch.
        if (!KnownTech.initialized)
            return;

        KnownTech.Deinitialize();
        KnownTech.Initialize(UnityEngine.Object.Instantiate<PDAData>(Player.main.pdaData));
    }
```

Also yes, the Instantiate is necessary as that is actually how the game itself does it so as not to corrupt the default one on the player prefab.

I mean we could leave the null check just for extra safety but the KnownTech by default games code CANNOT be initialized without the Player as its Player.Awake that triggers these methods in the first place.

Metious replied:
1. The reason why I'm not calling the game's Initialize directly is because that method resets everything, including the existing unlocked list and whatnot. So if we do call `KnownTech.Initialize`, the world will be left with only modded entries or none. I think we could probably work around this, but I thought leaving it as-is would be harmless. It's worth mentioning that since the UnknownName mod is coupled on the `KnownTech.Initialize`, it's already (before this PR) not working properly as we still aren't calling the method directly in Nautilus for the aforementioned reasons.
2. Valid. I forgot that the data's defaultTech is a List and not a HashSet, so even though they'll eventually _not_ be duplicates, the initial collection will grow and potentially cause some ridiculous lags.

> Also yes, the Instantiate is necessary as that is actually how the game itself does it so as not to corrupt the default one on the player prefab.

The game doesn't instantiate the PDAData at all and interact with it directly. It's harmless for us to also use it directly as long as we're not stacking changes over and over again.

Overall very good points and I will consider them.

MrPurple6411 replied:
> The game doesn't instantiate the PDAData at all and interact with it directly. It's harmless for us to also use it directly as long as we're not stacking changes over and over again.

![image](https://github.com/SubnauticaModding/Nautilus/assets/7858500/2ab25379-ee58-4a90-a35a-6ed6532f7f11)

MrPurple6411 replied:
All the modifications done in these initialize methods are done to a copy :-)

Metious replied:
Well, didn't know that. Interesting

Metious committed:
Compatibility fixes

We now call the original `KnownTech.Initialize` method to ensure mods that patch the method don't get unnoticed upon repatching.

Metious committed:
Add to the proper `knownCompoundTech` collection

MrPurple6411 reviewed (approved):
It looks right and doesn't seem to break any of my mods from the small testing I did.

And while my gut still feels like somethings wrong here it seems to work, and I can't figure out what my gut is trying to tell me so.......

Metious committed:
Fixed log spams for buildables

Since buildables are never registered to a crafting tree, Nautilus always logs a warning for that interpreting it as a mistake, when in reality it isn't.

Metious committed:
Added better error for requirements

MrPurple6411 commented on Nautilus/Assets/Gadgets/CraftingGadget.cs:
```diff
@@ -103,6 +100,11 @@ protected internal override void Build()
+        }
+        else if (!prefab.TryGetGadget(out ScanningGadget scanningGadget) || !scanningGadget.IsBuildable)
+        {
+            InternalLogger.Log($"Prefab '{prefab.Info.ClassID}' was not automatically registered into a crafting tree.");
```
Think we could make this Debug log instead?

Metious replied:
I think marking that as a debug log will make it easy to miss because a craftable item without a crafting tree that isn't a buildable is alarming and it's safer to assume that the developer missed something. In fact, I'd even argue it should be a warning instead.

MrPurple6411 reviewed (approved):
either way I think this is good enough. 

Just wondering about setting that 1 log to Debug.   if not then good, if yes then good, either way get this merged already lol

Metious committed:
Added hard-locking mechanism

Metious committed:
Added an overload for WithAnalysisTech

Added an overload for `WithAnalysisTech` that doesn't take a story goal. This way it's easier to write the same code for both SN and BZ.

MrPurple6411 reviewed (approved):
LGTM
